File: android/app/src/main/assets/debug/project/book.js
const {Collection} = require('./collection');
const {LZString} = require('./lzstring');

class BookCollection extends Collection {

    async fetch(url) {
        let pageUrl = new PageURL(url);

        let doc = await super.fetch(url);
        let lists = doc.querySelectorAll('.chapter-list'), results = [];
        if (lists.length === 0) {
            let stateNode = doc.querySelector('#__VIEWSTATE');
            if (stateNode) {
                let ctx = glib.ScriptContext.new('v8');
                ctx.eval(LZString);
                let html = ctx.eval('LZString.decompressFromBase64("' + stateNode.attr('value') + '")');
                doc = glib.GumboNode.parse2(html);
                lists = doc.querySelectorAll('.chapter-list');
            }
        }
        for (let list of lists) {
            let ul_arr = list.querySelectorAll('ul').reverse();
            for (let ul of ul_arr) {
                let li_arr = ul.querySelectorAll('li > a');
                for (let li of li_arr) {
                    let item = glib.DataItem.new();
                    item.type = glib.DataItem.Type.Chapter;
                    item.title = li.attr('title');
                    console.log(`title ${item.title}`);
                    item.link = pageUrl.href(li.attr('href'));
                    item.subtitle = li.querySelector('i').text;
                    results.push(item);
                }
            }
        }
        return results;
    }

    reload(_, cb) {
        console.log("reload");
        this.fetch(this.url).then((results) => {
            this.setData(results);
            cb.apply(null);
        }).catch(function(err) {
            if (err instanceof Error) 
                err = glib.Error.new(305, err.message);
            cb.apply(err);
        });
        return true;
    }
}

module.exports = function(data) {
    return BookCollection.new(data);
};

File: android/app/src/main/assets/debug/project/book.rb

class BookCollection < GS::Collection

  def initialize data
    @url = data.link
  end

  def fetch url
    req = GS::Request.create 'GET', url
    @callback = GS::Callback.block do 
      if req.has_error
        yield GS::Error.create(302, "Request error " + req.getError) 
      else
        body = req.getResponseBody
        if body
          yield nil, GS::GumboNode.parse(body)
        else
          yield GS::Error.create(301, "Response null body")
        end
      end
    end
    req.setOnComplete @callback
    req.start
  end

  def reload _, cb
    purl = PageURL.new @url
    fetch @url do |error, doc|
      if error
        cb.call error
      else
        info_data.summary = doc.querySelector('.detail-desc').text
        info_data.subtitle = doc.querySelectorAll('.detail-main-info-author a').map{|a| a.text}.join(',')
        links = doc.querySelectorAll '#tempc > ul > li > a'
        result = []
        links.each do |link|
          item = GS::DataItem.create 
          item.link = purl.href link.attr('href')
          title = link.querySelector('.detail-list-2-info-title')
          if title 
            item.title = title.text.gsub(/ +/, ' ') 
          else
            item.title = link.text.gsub(/ +/, ' ')
          end
          if link.querySelector('.detail-list-2-info-right')
            item.subtitle = 'VIP'
          end
          result << item
        end
        setData result
        cb.call nil
      end
    end
    true
  end

end

$exports = GS::Callback.block do |data| 
  BookCollection.create data
end